src/mpid/ch4/shm/xpmem/xpmem_control.h

```diff
@@ -0,0 +1,154 @@
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_SEND_LMT_ACK_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_ctrl_send_lmt_req_cb
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_req_t *slmt_req_hdr = &ctrl_hdr->xpmem_slmt_req;
+    MPIR_Request *rreq = NULL;
+    MPIR_Comm *root_comm;
+    MPIR_Request *anysource_partner;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_CTRL_SEND_LMT_REQ_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_CTRL_SEND_LMT_REQ_CB);
+
+    XPMEM_PT2PT_DBG_PRINT
+        ("send_lmt_req_cb: received src_offset 0x%lx, data_sz 0x%lx, sreq_ptr 0x%lx, "
+         "src_lrank %d, match info[src_rank %d, tag %d, context_id 0x%x]\n",
+         slmt_req_hdr->src_offset, slmt_req_hdr->data_sz, slmt_req_hdr->sreq_ptr,
+         slmt_req_hdr->src_lrank, slmt_req_hdr->src_rank, slmt_req_hdr->tag,
+         slmt_req_hdr->context_id);
+
+    /* Try to match a posted receive request. */
+    root_comm = MPIDIG_context_id_to_comm(slmt_req_hdr->context_id);
+    if (root_comm) {
+        /* MPIDI_CS_ENTER(); */
+        int continue_matching = 1;
+        while (continue_matching) {
+            anysource_partner = NULL;
+
+            rreq = MPIDIG_dequeue_posted(slmt_req_hdr->src_rank, slmt_req_hdr->tag,
+                                         slmt_req_hdr->context_id,
+                                         &MPIDIG_COMM(root_comm, posted_list));
+
+            if (rreq && MPIDI_REQUEST_ANYSOURCE_PARTNER(rreq)) {
+                /* Try to cancel NM parter request */
+                anysource_partner = MPIDI_REQUEST_ANYSOURCE_PARTNER(rreq);
+                mpi_errno = MPIDI_anysource_matched(anysource_partner,
+                                                    MPIDI_SHM, &continue_matching);
+                if (MPI_SUCCESS != mpi_errno)
+                    MPIR_ERR_POP(mpi_errno);
+
+                if (continue_matching) {
+                    /* NM partner request has already been matched, we need to continue until
+                     * no matching rreq. This SHM rreq will be cancelled by NM. */
+                    MPIR_Comm_release(root_comm);       /* -1 for posted_list */
+                    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
+                    continue;
+                }
+
+                /* Release cancelled NM partner request (only SHM request is returned to user) */
+                MPIDI_REQUEST_ANYSOURCE_PARTNER(rreq) = NULL;
+                MPIDI_REQUEST_ANYSOURCE_PARTNER(anysource_partner) = NULL;
+                MPIR_Request_free(anysource_partner);
+            }
+            break;
+        }
+        /* MPIDI_CS_EXIT(); */
+    }
+
+    /* Enqueue unexpected receive request */
+    if (rreq == NULL) {
```

Comment:
I think it makes more sense to swap the two cases here so the matched queue handling comes first (to go with the code above). Otherwise, you're bouncing back and forth between the cases.

Reply:
Swapped. I actually do not really like the code here as similar code exists also in `MPIDIG_send_target_msg_cb` and `MPIDIG_send_target_msg_cb`. But I think it is better to me to leave the the code refactoring in a separate PR to simplify this PR.